src/XCCDF_POLICY/xccdf_policy_remediate.c

```diff
@@ -1286,7 +1395,204 @@ static int _xccdf_policy_generate_fix_other(struct oscap_list *rules_to_fix, str
+{
+	struct oscap_iterator *service_disable_it = oscap_iterator_new(cmds->service_disable);
+	struct oscap_iterator *service_enable_it = oscap_iterator_new(cmds->service_enable);
+	if (oscap_iterator_has_more(service_disable_it) || oscap_iterator_has_more(service_enable_it)) {
+		_write_text_to_fd(output_fd, "# Disable and enable systemd services based on the SCAP profile\n");
+		_write_text_to_fd(output_fd, "services");
+		if (oscap_iterator_has_more(service_disable_it)) {
+			_write_text_to_fd(output_fd, " --disabled=");
+			while (oscap_iterator_has_more(service_disable_it)) {
+				char *command = (char *) oscap_iterator_next(service_disable_it);
+				_write_text_to_fd(output_fd, command);
+				if (oscap_iterator_has_more(service_disable_it))
+					_write_text_to_fd(output_fd, ",");
+			}
+		}
+		if (oscap_iterator_has_more(service_enable_it)) {
+			_write_text_to_fd(output_fd, " --enabled=");
+			while (oscap_iterator_has_more(service_enable_it)) {
+				char *command = (char *) oscap_iterator_next(service_enable_it);
+				_write_text_to_fd(output_fd, command);
+				if (oscap_iterator_has_more(service_enable_it))
+					_write_text_to_fd(output_fd, ",");
+			}
+		}
+		_write_text_to_fd(output_fd, "\n\n");
+	}
+	oscap_iterator_free(service_disable_it);
+	oscap_iterator_free(service_enable_it);
+	return 0;
+}
+
+static int _generate_kickstart_packages(struct kickstart_commands *cmds, int output_fd)
+{
+	_write_text_to_fd(output_fd, "# Packages selection (%packages section is required)\n");
+	_write_text_to_fd(output_fd, "%packages\n");
+	/* openscap-scanner and scap-security-guide needs to be installed because we will run oscap in the %post section */
+	_write_text_to_fd(output_fd, "openscap-scanner\n");
+	_write_text_to_fd(output_fd, "scap-security-guide\n");
+	struct oscap_iterator *package_install_it = oscap_iterator_new(cmds->package_install);
+	while (oscap_iterator_has_more(package_install_it)) {
+		char *package = (char *) oscap_iterator_next(package_install_it);
+		_write_text_to_fd(output_fd, package);
+		_write_text_to_fd(output_fd, "\n");
+	}
+	oscap_iterator_free(package_install_it);
+	struct oscap_iterator *package_remove_it = oscap_iterator_new(cmds->package_remove);
+	while (oscap_iterator_has_more(package_remove_it)) {
+		char *package = (char *) oscap_iterator_next(package_remove_it);
+		_write_text_to_fd(output_fd, "-");
+		_write_text_to_fd(output_fd, package);
+		_write_text_to_fd(output_fd, "\n");
+	}
+	oscap_iterator_free(package_remove_it);
+	_write_text_to_fd(output_fd, "%end\n");
+	_write_text_to_fd(output_fd, "\n");
+	return 0;
+}
+
+static int _generate_kickstart_post(struct kickstart_commands *cmds, const char *profile_id, const char *input_path, int output_fd)
+{
+	_write_text_to_fd(output_fd, "%post\n");
+	char *dup = strdup(input_path);
+	char *basename = oscap_basename(dup);
+	free(dup);
+	char *oscap_command = oscap_sprintf(
+		"oscap xccdf eval --remediate --profile '%s' /usr/share/xml/scap/ssg/content/%s\n",
```

Comment:
This will work sometimes, but sometimes it won't work. How do we ensure that the file is the same as the file that the user provided to `oscap` when generating the kickstart? And even a bigger problem, what if the user used a tailoring when generating the kickstart?

Reply:
This is why we need to at least fix this: https://github.com/OpenSCAP/openscap/issues/771.

Reply:
Amazing! I didn't know that we have issue for that. Great news is that the issue is already "accidentally" fixed in this PR! The headers now contain actual file path of the file used as input to the `oscap xccdf generate fix` command.

The knowledge of the actual input file path is also reused in this place. Here, I take its base name and concatenate it with the well-known scap-security-guide directory path. This way I achieved that it will work sometimes. The sometimes specifically means that 1. user uses a data stream from the scap-security-guide package to generate the KS and 2. the version of the scap-security-guide package used to generate the KS is the same that will be installed on the installed system and 3. user doesn't use tailoring when generating the KS. If any of the three conditions isn't true then it won't work or will give surprising results. I think that's insufficient therefore I opened this comment.

Reply:
I have this proposal how to solve this problem. We will document that the recommended workflow is to serve the SCAP source data stream and the optional XCCDF tailoring file on a web server. Then, the generated kickstart will contain a command to download these files from the server. We will not install scap-security-guide package during the installation. This way it will make sure that the same files are used for both kickstart generation before installation and remediation during installation. The generated kickstart can contain placeholders for the content URI and comments highlighting it. Also, we can have a CLI option that would get propagated to the generated kickstart. The key thing would be to have the workflow well documented.

Reply:
Another proposal has been suggested by @dahaic : The kickstart would explicitly request the specific version of scap-security-guide RPM package to be installed. The package would be fetch from CDN. This way we would ensure that the same data stream is used for the remediation as is used for the kickstart generation. The tailoring file could be inserted in a form of base64 blob or in a form of autotailor command.

Reply:
1. It'd be nice to actually have just a full path here for the content. We don't care much about non-packaged DSes. It's up to the advanced customer to figure out how they want to fetch, store and use it (we can give an example in the docs, though).
2. If we can somehow embed tailoring (when `--tailoring` option is provided along with `generate fix`) and generate proper KS that would use it without requiring the user to serve it separately — it would be very nice, but let's not concentrate on it before everything else is complete.

Reply:
If we could be able to bring tailoring into the remediation generator context it will help us to also generate self-contained tailored Blueprints.